### PR TITLE
Work around Visual Studio bug inferring type of auto

### PR DESCRIPTION
**Pull request by hughbe**
https://connect.microsoft.com/VisualStudio/feedback/details/3121204/visual-studio-reports-cannot-deduce-auto-type-for-an-auto-despite-msvc-successfully-compiling-the-code

Although ProtocolConformance.cpp is part of the runtime, so must be built with Clang, Visual Studio (not MSVC) reports "cannot infer type of auto" errors, severely limiting it's ability to provide a good IDE experience for any compiler.

### CODE REVIEW

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by hughbe** [comment]
@swift-ci please smoke test

File: include/swift/Runtime/Config.h [hunk 1]
```diff
@@ -140,6 +140,10 @@
 // been integrated into clang and llvm.
 #define SWIFT_RT_USE_RegisterPreservingCC 0
 
+#if !defined(__has_attribute)
+#define __has_attribute(x) 0
```
Comment (gparker42) [inline include/swift/Runtime/Config.h]: We already have this in two places. It should be centralized somewhere.

File: stdlib/public/runtime/ProtocolConformance.cpp [hunk 2]
```diff
@@ -373,7 +373,7 @@ searchInConformanceCache(const Metadata *type,
     // For generic and resilient types, nondependent conformances
     // are keyed by the nominal type descriptor rather than the
     // metadata, so try that.
-    auto *description = type->getNominalTypeDescriptor().get();
+    auto description = type->getNominalTypeDescriptor().get();
```
Comment (gparker42) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Does `const auto *` work for these? I think we want to avoid downgrading qualifiers on `auto`.

Comment (hughbe) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Unfortunately, `const auto*` does not work. `const auto` does however
untitled

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by hughbe** [comment]
@swift-ci please smoke test

File: include/swift/Runtime/Config.h (hunk 1 above)
Comment (hughbe) [inline include/swift/Runtime/Config.h]: Yeah true. What I've done is delegated that kind of stuff to "LLVM/Support/Compiler.h" - see the 3rd commit

**Comment by hughbe** [comment]
@gparker42 you happy with this? I think `const auto` is the best we can do here

**Comment by hughbe** [comment]
Friendly ping, should be a simple set of changes :)

**Comment by gparker42** [comment]
Looks good. `const auto` is a better disambiguator than `auto *` IMO.

**Pull request merged**